events/renderers.py

```diff
@@ -12,3 +26,198 @@ class JSONLDRenderer(JSONRenderer):
+        if start_date:
+            dates['start'] = dateutil.parser.parse(start_date)
+
+        end_date = self.get_query_params(renderer_context, 'end')
+        if end_date:
+            dates['end'] = dateutil.parser.parse(end_date)
+
+        try:
+            days = int(self.get_query_params(renderer_context, 'days'))
+        except ValueError as e:
+            days = None
+        # Days override dates
+        if days:
+            dates['start'] = datetime.datetime.now().date()
+            dates['end'] = dates['start'] + datetime.timedelta(days=days)
+
+        return dates
+
+    def render(self, data, media_type=None, renderer_context=None):
+        language_code = getattr(settings, 'LANGUAGE_CODE', 'fi')
+        document = Document()
+
+        location = self.get_query_params(renderer_context, 'location')
+        # This could be set to something like "All locations"
+        place_name = ''
+        try:
+            place_name = Place.objects.values_list(
+                'name',
+                flat=True,
+            ).get(id=location)
+        except Place.DoesNotExist:
+            raise Http404('Location not found')
+
+        document.add_heading('Tapahtumat: %s' % place_name, 0)
+
+        # Note that these dates are for the request while the dates for
+        # individual events differ (but should be within these).
+        dates = self.get_date_range(renderer_context)
+
+        # These dates are for the entire document, not individual events.
+        # The minuses in strftime are for removing leading zeros.
+        date_string = ''
+        if dates['start']:
+            date_string += dates['start'].strftime('%-d.%-m.%Y')
+        if dates['end']:
+            date_string += '-%s' % dates['end'].strftime('%-d.%-m.%Y')
+
+        if date_string:
+            document.add_paragraph(date_string)
+
+        # We need to sort the events in chronological order, so that events
+        # that have differing starting and ending dates don't end up in the
+        # middle of events that are on a single day.
+        events = []
+        for event in sorted(data['data'], key=lambda e: e['start_time_obj']):
+            events.append(event)
+
+        # If no start date is found, parse it from whatever the first event is
+        if not dates['start']:
+            dates['start'] = dateutil.parser.parse(events[0]['start_time'])
+
+        # If no end date exists, get the last event and parse the end date
+        # from the last event.
+        if not dates['end']:
+            last_event = sorted(
+                data['data'],
+                key=lambda e: e['end_time_obj']
+            )[-1]
+            dates['end'] = dateutil.parser.parse(last_event['end_time'])
+
+        if not events:
+            raise Http404('No events for location')
+
+        current_start_day = None
+
+        for event in events:
+            event_name = self.get_any_language(
```

Comment:
This loop body could be hoisted into a method of its own (that processes a single event).